src/sentry/grouping/variants.py

```diff
@@ -241,10 +245,11 @@ def get_hash(self) -> str | None:
                 final_values.append(value)
         return hash_from_values(final_values)
 
-    def _get_metadata_as_dict(self):
-        rv = ComponentVariant._get_metadata_as_dict(self)
-        rv.update(expose_fingerprint_dict(self.values, self.info))
-        return rv
+    def _get_metadata_as_dict(self) -> Mapping[str, Any]:
```

Comment:
would be nice if this were a TypedDict but still a slight improvement from before

this should actually return `-> dict[str, Any]` however -- rule of thumb is to always return exact, concrete types

Reply:
Check out [the note in the base class](https://github.com/getsentry/sentry/pull/82206/files#diff-dd17c2a5c5dc7a78219fadcf3aff212fa9010bb566c2e2e55884a6faea07ce5fR43-R45):

```python
    # This has to return `Mapping` rather than `dict` so that subtypes can override the return value
    # with a TypedDict if they choose. See https://github.com/python/mypy/issues/4976.
    def _get_metadata_as_dict(self) -> Mapping[str, Any]:
```

At least one subclass does already return a `TypedDict`.

Reply:
I'm going to merge this as is for now, since I can't make the `dict`/`Mapping` swap, but will put on my todo list to think about making typed dicts for each subclass as a further improvement.